kernel/src/actions/visitors.rs

```diff
@@ -591,6 +518,87 @@ pub(crate) fn visit_deletion_vector_at<'a>(
+    row_index: usize,
+    getters: &[&'a dyn GetData<'a>],
+) -> DeltaResult<Option<Metadata>> {
+    require!(
+        getters.len() == 9,
+        Error::InternalError(format!(
+            "Wrong number of MetadataVisitor getters: {}",
+            getters.len()
+        ))
+    );
+    // Since id column is required, use it to detect presence of a metadata action
+    if let Some(id) = getters[0].get_opt(row_index, "metadata.id")? {
+        let name: Option<String> = getters[1].get_opt(row_index, "metadata.name")?;
+        let description: Option<String> = getters[2].get_opt(row_index, "metadata.description")?;
+        // get format out of primitives
+        let format_provider: String = getters[3].get(row_index, "metadata.format.provider")?;
+        // options for format is always empty, so skip getters[4]
+        let schema_string: String = getters[5].get(row_index, "metadata.schema_string")?;
+        let partition_columns: Vec<_> = getters[6].get(row_index, "metadata.partition_list")?;
+        let created_time: Option<i64> = getters[7].get_opt(row_index, "metadata.created_time")?;
+        let configuration_map_opt: Option<HashMap<_, _>> =
+            getters[8].get_opt(row_index, "metadata.configuration")?;
+        let configuration = configuration_map_opt.unwrap_or_else(HashMap::new);
+
+        Ok(Some(Metadata {
+            id,
+            name,
+            description,
+            format: Format {
+                provider: format_provider,
+                options: HashMap::new(),
+            },
+            schema_string,
+            partition_columns,
+            created_time,
+            configuration,
+        }))
+    } else {
+        Ok(None)
+    }
+}
+
+/// Get a Protocol out of some engine data. The caller is responsible for slicing the `getters` slice
+/// such that the first element contains the `min_reader_version` element of the protocol.
+#[internal_api]
+pub(crate) fn visit_protocol_at<'a>(
+    row_index: usize,
+    getters: &[&'a dyn GetData<'a>],
+) -> DeltaResult<Option<Protocol>> {
+    require!(
+        getters.len() == 4,
+        Error::InternalError(format!(
+            "Wrong number of ProtocolVisitor getters: {}",
+            getters.len()
+        ))
+    );
+    // Since minReaderVersion column is required, use it to detect presence of a Protocol action
+    if let Some(min_reader_version) =
+        getters[0].get_opt(row_index, "protocol.min_reader_version")?
+    {
+        let min_writer_version: i32 = getters[1].get(row_index, "protocol.min_writer_version")?;
+        let reader_features: Option<Vec<_>> =
+            getters[2].get_opt(row_index, "protocol.reader_features")?;
+        let writer_features: Option<Vec<_>> =
+            getters[3].get_opt(row_index, "protocol.writer_features")?;
+
+        Ok(Some(Protocol::try_new(
+            min_reader_version,
+            min_writer_version,
+            reader_features,
+            writer_features,
+        )?))
```

Comment:
```suggestion
        Some(Protocol::try_new(
            min_reader_version,
            min_writer_version,
            reader_features,
            writer_features,
        )).transpose()
```

Reply:
Or, if the `?` is confusing, maybe just admit they're two operations and split them up?
```suggestion
        let protocol = Protocol::try_new(
            min_reader_version,
            min_writer_version,
            reader_features,
            writer_features,
        )?;
        Ok(Some(protocol))
```
(shorter isn't _always_ better... just usually :P )